### ISSUE TITLE
nsenter: give capabilities to a process

### ISSUE THREAD

**Issue by Sacred-Salamander** [util-linux/util-linux#1946]
I have an existing namespace from this
```sh
bwrap --unshare-user --unshare-pid --dev-bind / / --proc /proc --tmpfs /run --tmpfs /tmp --cap-add CAP_SYS_ADMIN /bin/bash
```

which I'm trying to enter with nsenter using this
```sh
nsenter --preserve-credentials --user --mount --pid --wdns="${HOME}" -t <pid_of_bwrap> /bin/bash
```

I can now check the capabilites for the second shell
```sh
grep Cap /proc/self/status
```

```text
CapInh:	0000000000000000
CapPrm:	0000000000000000
CapEff:	0000000000000000
CapBnd:	000001ffffffffff
CapAmb:	0000000000000000
```

**Comment by Sacred-Salamander** [comment]
I'm sorry, I should have been more clear, I'm trying to mount a fuse fs inside my namespace from the second terminal

It works from the first terminal only if I use bubblewrap with the --cap-add CAP_SYS_ADMIN option, I thought that was needed, but on the host I don't need them to use fuse mounts as a user. 

It might be that I still need it for fuse inside namespaces, but I'm not sure exactly what I need to make it work when I enter the namespace through nsenter

**Comment by karelzak** [comment]
I was able to reproduce your problem two weeks ago, but I had not time to investigate it.

**Comment by brauner** [comment]
> I'm sorry, I should have been more clear, I'm trying to mount a fuse fs inside my namespace from the second terminal
> 
> It works from the first terminal only if I use bubblewrap with the --cap-add CAP_SYS_ADMIN option, I thought that was needed, but on the host I don't need them to use fuse mounts as a user.
> 
> It might be that I still need it for fuse inside namespaces, but I'm not sure exactly what I need to make it work when I enter the namespace through nsenter

Whether or not FUSE can be mounted from inside a user namespace depends on the kernel version. The fact that you can mount FUSE on the host doesn't necessarily mean much. The `fusermount` binary is often `setuid`.

**Comment by Sacred-Salamander** [comment]
Thanks both of you for helping

The fusermount is indeed setuid

FUSE can be mounted inside the user namespace, if I use bubblewrap with CAP_SYS_ADMIN, so I don't think it's the kernel version

When I join the same namespaces with nsenter everything seems to be fine but I can't use the fuse mount inside the second shell spawned with nsenter, the first one works

If nsenter allowed me to set capabilities, would it work then you think? Could there be any other workaround for doing this from the shell started by nsenter?

**Comment by brauner** [comment]
I see. So `bwrap` does give you regular user, e.g., `1000`, `CAP_SYS_ADMIN`. Then it can of course mount `FUSE` inside of user namespaces (given the correct kernel version ofc). But when you use `nsenter` from what I see here you don't have `CAP_SYS_ADMIN`. So yes, if `nsenter` where to give you that it would work.

**Comment by Sacred-Salamander** [comment]
Okey thank you

I guess for now nsenter can't do this, and that there aren't any workarounds for inheriting another's process capabilities either

Does nsenter drop all capabilites? Or shouldn't it inherit the parents capabilites otherwise? 

I hope it can be made an option for nsenter in the future, this should be up to the caller to decide

Maybe an option for inheriting the current capabilities of the parent

Maybe there is more trickery to get this to work, but I would be glad to help with anything to debug this

**Comment by karelzak** [comment]
BTW, there is a feature request from Lennart to add an option to enter cgroup or the  target (PID): https://github.com/util-linux/util-linux/issues/2006

**Comment by Sacred-Salamander** [comment]
I can confirm that the newly added option `-c, --join-cgroup` solves my problem

My cgroup.procs had wrong ownership for some reason, was assigned to root:root, so it wouldn't work until I changed it with
`chown 1000:1000 /sys/fs/group//user.slice/user-1000.slice/session-3.scope/cgroup.procs` as root

**Comment by Sacred-Salamander** [comment]
Sorry I tricked myself, I never got into the namespaces or cgroup when I run nsenter with --join-cgroup, I thought I did since the shell looks the same and I do get into a new shell when I run without the option, and the mount shows up in my original namespaced shell because I executed the mount outside any namespace, so I thought it worked, no error or exit code was thrown but when inspecting with strace I can see it just closes the process

```txt
$ strace nsenter --preserve-credentials --user --mount --pid --wdns="${HOME}" --join-cgroup -t 1377433 /bin/bash

statfs("/sys/fs/cgroup", {f_type=CGROUP2_SUPER_MAGIC, f_bsize=4096, f_blocks=0, f_bfree=0, f_bavail=0, f_files=0, f_ffree=0, f_fsid={val=[0, 0]}, f_namelen=255, f_frsize=4096, f_flags=ST_VALID|ST_NOSUID|ST_NODEV|ST_NOEXEC|ST_RELATIME}) = 0
close(0)                                = 0
openat(AT_FDCWD, "/proc/1377433/cgroup", O_RDONLY) = 0
read(0, "0::/user.slice/user-1000.slice/s"..., 1024) = 47
read(0, "", 977)                        = 0
read(0, "", 977)                        = 0
openat(AT_FDCWD, "/sys/fs/cgroup//user.slice/user-1000.slice/session-3.scope/cgroup.procs", O_WRONLY|O_APPEND) = 6
setns(4, CLONE_NEWPID)                  = -1 EPERM (Operation not permitted)
setns(5, CLONE_NEWNS)                   = -1 EPERM (Operation not permitted)
setns(3, CLONE_NEWUSER)                 = 0
close(3)                                = 0
setns(4, CLONE_NEWPID)                  = 0
close(4)                                = 0
setns(5, CLONE_NEWNS)                   = 0
close(5)                                = 0
openat(AT_FDCWD, "/home/user", O_RDONLY) = 3
fchdir(3)                               = 0
close(3)                                = 0
getpid()                                = 1399296
write(6, "1399296", 7)                  = 7
rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[CHLD], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f57c4169520}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f57c412ea10) = 1399297
wait4(1399297, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], WSTOPPED, NULL) = 1399297
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1399297, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
dup(1)                                  = 3
close(3)                                = 0
dup(2)                                  = 3
close(3)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```


Running without the option it would freeze at the wait4 call until the shell exited
```
wait4(1399297, 
```

**Comment by Sacred-Salamander** [comment]
This time I can confirm the newly added option `--keep-caps` solves my problem instead
It is available in nsenter >= 2.40
No `-c, --join-cgroup` required for me either
